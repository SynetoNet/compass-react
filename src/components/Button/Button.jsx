import React, { useEffect, useState, useRef } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Ink from "react-ink";

import BootstrapButton from "react-bootstrap/Button";

import "./Button.scss";
import "../../base/fixed-width.scss";

const LOADING_TEXTS = [
  "Still working...",
  "Mostly done...",
  "A bit more...",
  "Almost done..."
];

const LOADING_DELAY = 10000;

const Button = ({
  appearance,
  width,
  loading,
  disabled,
  block,
  children,
  onClick,
  role,
  className,
  ...props
}) => {
  const initialText = children;
  const [text, setText] = useState(initialText);
  const [loadingIndex, setLoadingIndex] = useState(-1);
  const prevLoading = usePrevious(loading);
  const _disabled = loading || disabled;

  let classes = classNames(className, {
    ["width-" + width]: width,
    disabled: _disabled,
  });

  let buttonTexts = [];

  if (loading) {
    const initialLoadingText =
      typeof loading === "string" ? loading : "Please wait...";

    buttonTexts = [initialLoadingText, ...LOADING_TEXTS];
  }

  useEffect(() => {
    let timer;

    if (loading) {
      if (!prevLoading) {
        // switched from not loading, to loading
        updateLoadingText();
      }

      loadingIndex < buttonTexts.length - 1 &&
        (timer = setTimeout(() => updateLoadingText(), LOADING_DELAY));
    } else if (prevLoading) {
      reset();
    }

    return () => {
      clearTimeout(timer);
    };
  }, [loading, loadingIndex]);

  useEffect(() => {
    let newText = loading ? buttonTexts[loadingIndex] : children;
    setText(newText);
  }, [children, loadingIndex]);

  function updateLoadingText() {
    setLoadingIndex(prev => prev + 1);
  }

  function reset() {
    setText(initialText);
    setLoadingIndex(-1);
  }

  function _onClick(e) {
    !_disabled && onClick && onClick(e);
  }

  if (role === 'secondary') {
    appearance = "outline-" + appearance;
  }

  if (role === 'tertiary' || role === 'link') {
    classes += ' btn-' + appearance;
    appearance = "link";
  }

  const bootstrapBtn = (
    <BootstrapButton
      variant={appearance}
      className={classes}
      disabled={_disabled}
      onClick={_onClick}
      {...props}
    >
      {text}
      <Ink />
    </BootstrapButton>
  );

  if (!block) {
    return bootstrapBtn;
  }
  
  return <div className="d-grid">{bootstrapBtn}</div> 
};

Button.propTypes = {
  role: PropTypes.oneOf(["primary", "tertiary", "secondary", "link"]),
  appearance: PropTypes.oneOf(["primary", "success", "danger", "warning", "info"]),
  width: PropTypes.oneOf([32, 64, 96, 128, "100%"]),
  type: PropTypes.oneOf(["button", "submit", "reset"]),
  disabled: PropTypes.bool,
  block: PropTypes.bool,
  loading: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  children: PropTypes.node,
  onClick: PropTypes.func
};

Button.defaultProps = {
  role: "primary",
  appearance: "primary",
  type: "button",
  width: "100%",
  disabled: false,
  block: false
};

function usePrevious(value) {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}

export default Button;
