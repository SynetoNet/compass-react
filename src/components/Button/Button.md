#### **Default Buttons**

```jsx
<Button>Default Button</Button>
<i className="pr-2" />
<Button color="success">Success</Button>
<i className="pr-2" />
<Button color="danger">Danger</Button>
<i className="pr-2" />
<Button color="warning">Warning</Button>
<i className="pr-2" />
<Button color="info">Info</Button>
```

#### **Primary Buttons**

```jsx
<Button variant="primary">Primary</Button>
<i className="pr-2" />
<Button variant="primary" color="success">Success</Button>
<i className="pr-2" />
<Button variant="primary" color="danger">Danger</Button>
<i className="pr-2" />
<Button variant="primary" color="warning">Warning</Button>
<i className="pr-2" />
<Button variant="primary" color="info">Info</Button>
```

#### **Tertiary Buttons**

```jsx
<Button variant="tertiary">Tertiary</Button>
<i className="pr-2" />
<Button variant="tertiary" color="success">Success</Button>
<i className="pr-2" />
<Button variant="tertiary" color="danger">Danger</Button>
<i className="pr-2" />
<Button variant="tertiary" color="warning">Warning</Button>
<i className="pr-2" />
<Button variant="tertiary" color="info">Info</Button>
```

#### **Link**

```jsx
<Button variant="link" href="https://github.com/SynetoNet/compass-react">
  Link
</Button>
```

#### **Disabled Buttons**

```jsx
<div>
  <Button disabled>Disabled</Button>
  <i className="pr-2" />
  <Button disabled color="success">Success</Button>
  <i className="pr-2" />
  <Button disabled color="danger">Danger</Button>
  <i className="pr-2" />
  <Button disabled color="warning">Warning</Button>
  <i className="pr-2" />
  <Button disabled color="info">Info</Button>
</div>

<div className="mt-2">
  <Button disabled variant="primary">Disabled</Button>
  <i className="pr-2" />
  <Button disabled variant="primary" color="success">Success</Button>
  <i className="pr-2" />
  <Button disabled variant="primary" color="danger">Danger</Button>
  <i className="pr-2" />
  <Button disabled variant="primary" color="warning">Warning</Button>
  <i className="pr-2" />
  <Button disabled variant="primary" color="info">Info</Button>
</div>

<div className="mt-2">
  <Button disabled variant="tertiary">Disabled</Button>
  <i className="pr-2" />
  <Button disabled variant="tertiary" color="success">Success</Button>
  <i className="pr-2" />
  <Button disabled variant="tertiary" color="danger">Danger</Button>
  <i className="pr-2" />
  <Button disabled variant="tertiary" color="warning">Warning</Button>
  <i className="pr-2" />
  <Button disabled variant="tertiary" color="info">Info</Button>
</div>
```

#### **Modular width**

```jsx
<Button width={32}>32px</Button>
<div className="pt-3" />
<Button width={64}>64px</Button>
<div className="pt-3" />
<Button width={96}>96px</Button>
<div className="pt-3" />
<Button width={128}>128px</Button>
```

#### **Loading Button controlled**

To handle Loading state, you could manually control the:

- disabled state
- onClick handler
- Button text

```jsx
import { useState } from "react";

const LoadingExample = () => {
  const [isLoading, toggleLoading] = useState(false);

  const handleClick = () => {
    toggleLoading(true);
    setTimeout(() => toggleLoading(false), 2000);
  };

  return (
    <Button
      variant="primary"
      disabled={isLoading}
      onClick={!isLoading ? handleClick : null}
    >
      {isLoading ? "Please wait 2s" : "Click to load"}
    </Button>
  );
};

<LoadingExample />;
```

#### **Loading Button uncontrolled**

For easier usage and for long running actions, you should use the **`loading`** prop:

- defaults to "Please wait..." when passing true;
- cycles through various messages every 10s;

```jsx
import { useState } from "react";

const LoadingExample = () => {
  const [loading, setLoading] = useState(false);

  const handleClick = () => {
    setLoading("Hold on");
    setTimeout(() => setLoading(false), 2500);
  };

  return (
    <Button variant="primary" loading={loading} onClick={handleClick}>
      {"Click to load"}
    </Button>
  );
};

<LoadingExample />;
```
