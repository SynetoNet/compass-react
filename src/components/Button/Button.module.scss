@import "../../base/colors";

.btn {
  background: white;
  color: $accent;
  border-color: $accent;
  border-radius: 3px;
  text-align: center;
}
.btn-primary {
  background: $accent;
  color: white;
  border-color: $accent;
}
.btn:active,
.btn:focus,
.btn:hover {
  background: white;
  color: $accent;
  border-color: $accent;
}

.btn-secondary:not(:disabled):not(.disabled):active,
.btn-secondary:not(:disabled):not(.disabled).active,
.show > .btn-secondary.dropdown-toggle {
  background: white;
  color: $accent;
  border-color: $accent;
}

.btn:disabled {
  /* background-color: var(--color-accent); */
  color: var(--color-neutral-40);
  border-color: var(--color-neutral-40);
}
.btn:disabled {
  background-color: var(--color-neutral-6);
  border-color: var(--color-neutral-6);
}
