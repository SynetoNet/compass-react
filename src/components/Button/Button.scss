@import "../../base/colors";
@import "../../base/variables";

.btn {
  border-radius: 2px;
  position: relative;

  &.disabled,
  &:disabled {
    cursor: not-allowed;
    opacity: 0.5;
  }
}

.btn-primary {
  &:hover:not(:disabled) {
    background-color: $semantic-accent-dark-20;
  }
  &:active:not(:disabled) {
    background-color: $semantic-accent-dark-30;
  }

  &.btn-success {
    background-color: $semantic-confirm;
    &:hover:not(:disabled) {
      background-color: $semantic-confirm-dark-20;
    }
    &:active:not(:disabled) {
      background-color: $semantic-confirm-dark-30;
    }
  }
  &.btn-danger {
    background-color: $semantic-danger;
    &:hover:not(:disabled) {
      background-color: $semantic-danger-dark-20;
    }
    &:active:not(:disabled) {
      background-color: $semantic-danger-dark-30;
    }
  }
  &.btn-info {
    background-color: $semantic-info;
    &:hover:not(:disabled) {
      background-color: $semantic-info-dark-20;
    }
    &:active:not(:disabled) {
      background-color: $semantic-info-dark-30;
    }
  }
  &.btn-warning {
    background-color: $semantic-warn;
    color: $neutral-0;
    &:hover:not(:disabled) {
      background-color: $semantic-warn-dark-20;
    }
    &:active:not(:disabled) {
      background-color: $semantic-warn-dark-30;
    }
  }
}

// .btn-secondary:not(:disabled):not(.disabled):active,
// .btn-secondary:not(:disabled):not(.disabled).active,
.show > .btn-secondary.dropdown-toggle,
.btn-secondary {
  background-color: $neutral-0;
  border: 1px solid $semantic-accent-contrast;
  color: $semantic-accent-contrast;

  &:hover:not(:disabled) {
    background-color: $semantic-accent-dark-20;
    border-color: $semantic-accent-dark-20;
    color: $neutral-0;
  }
  &:active:not(:disabled) {
    background-color: $semantic-accent-dark-30 !important;
    border-color: $semantic-accent-dark-30 !important;
    color: $neutral-0;
  }
  &:focus {
    box-shadow: 0 0 0 0.2rem $semantic-accent-light-60 !important;
  }
  &:disabled {
    border-color: $semantic-accent-contrast;
    color: $semantic-accent-contrast;
    background-color: $neutral-0;
  }

  &.btn-success {
    background-color: $neutral-0;
    border-color: $semantic-confirm;
    color: $semantic-confirm-dark-30;

    &:hover:not(:disabled) {
      background-color: $semantic-confirm-dark-20;
      border-color: $semantic-confirm-dark-20;
      color: $neutral-0;
    }
    &:active:not(:disabled) {
      background-color: $semantic-confirm-dark-30 !important;
      border-color: $semantic-confirm-dark-30 !important;
      color: $neutral-0;
    }
    &:focus {
      box-shadow: 0 0 0 0.2rem $semantic-confirm-light-60 !important;
    }
  }

  &.btn-danger {
    background-color: $neutral-0;
    border-color: $semantic-danger;
    color: $semantic-danger-dark-30;

    &:hover:not(:disabled) {
      background-color: $semantic-danger-dark-20;
      border-color: $semantic-danger-dark-20;
      color: $neutral-0;
    }
    &:active:not(:disabled) {
      background-color: $semantic-danger-dark-30 !important;
      border-color: $semantic-danger-dark-30 !important;
      color: $neutral-0;
    }
    &:focus {
      box-shadow: 0 0 0 0.2rem $semantic-danger-light-60 !important;
    }
  }

  &.btn-warning {
    background-color: $neutral-0;
    border-color: $semantic-warn;
    color: $semantic-warn-dark-30;

    &:hover:not(:disabled) {
      background-color: $semantic-warn-dark-20;
      border-color: $semantic-warn-dark-20;
      color: $neutral-0;
    }
    &:active:not(:disabled) {
      background-color: $semantic-warn-dark-30 !important;
      border-color: $semantic-warn-dark-30 !important;
      color: $neutral-0;
    }
    &:focus {
      box-shadow: 0 0 0 0.2rem $semantic-warn-light-60 !important;
    }
  }

  &.btn-info {
    background-color: $neutral-0;
    border-color: $semantic-info;
    color: $semantic-info-dark-30;

    &:hover:not(:disabled) {
      background-color: $semantic-info-dark-20;
      border-color: $semantic-info-dark-20;
      color: $neutral-0;
    }
    &:active:not(:disabled) {
      background-color: $semantic-info-dark-30 !important;
      border-color: $semantic-info-dark-30 !important;
      color: $neutral-0;
    }
    &:focus {
      box-shadow: 0 0 0 0.2rem $semantic-info-light-60 !important;
    }
  }
}

.btn-tertiary {
  color: $semantic-accent-contrast;
  border-color: transparent !important;
  background: none;
  &:disabled {
    background: none;
  }
  &:hover:not(:disabled) {
    background-color: $semantic-accent-light-85;
    color: $semantic-accent-dark-30;
  }
  &:active:not(:disabled) {
    background-color: $semantic-accent-light-60;
    color: $semantic-accent-dark-30;
  }
  &:disabled {
    color: $semantic-accent-contrast;
  }

  &.btn-success {
    color: $semantic-confirm;
    background-color: $neutral-0;

    &:hover:not(:disabled) {
      background-color: $semantic-confirm-light-85;
      color: $semantic-confirm-dark-30;
    }
    &:active:not(:disabled) {
      background-color: $semantic-confirm-light-60 !important;
      color: $semantic-confirm-dark-30 !important;
    }
  }

  &.btn-danger {
    color: $semantic-danger;
    background-color: $neutral-0;

    &:hover:not(:disabled) {
      background-color: $semantic-danger-light-85;
      color: $semantic-danger-dark-30;
    }
    &:active:not(:disabled) {
      background-color: $semantic-danger-light-60 !important;
      color: $semantic-danger-dark-30 !important;
    }
  }

  &.btn-warning {
    color: $semantic-warn-dark-20;
    background-color: $neutral-0;

    &:hover:not(:disabled) {
      background-color: $semantic-warn-light-85;
      color: $semantic-warn-dark-30;
    }
    &:active:not(:disabled) {
      background-color: $semantic-warn-light-60 !important;
      color: $semantic-warn-dark-30 !important;
    }
  }

  &.btn-info {
    color: $semantic-info;
    background-color: $neutral-0;

    &:hover:not(:disabled) {
      background-color: $semantic-info-light-85;
      color: $semantic-info-dark-30;
    }
    &:active:not(:disabled) {
      background-color: $semantic-info-light-60 !important;
      color: $semantic-info-dark-30 !important;
    }
  }
}

.btn-link {
  background: none;
  border: none;
  &:hover:not(:disabled) {
    background-color: $semantic-accent-light-85;
    color: $semantic-accent-dark-30;
  }
  &:active:not(:disabled) {
    background-color: $semantic-accent-light-60;
    color: $semantic-accent-dark-30;
  }
  &:disabled {
    color: $semantic-accent-contrast;
  }
}