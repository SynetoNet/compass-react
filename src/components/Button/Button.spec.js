import React from "react";
import Button from "./Button";
import renderer from "react-test-renderer";

describe("Default Button", () => {
  test("secondary Button", () => {
    const component = renderer.create(<Button>Secondary</Button>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("primary Button", () => {
    const component = renderer.create(
      <Button role="primary">Primary</Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("tertiary Button", () => {
    const component = renderer.create(
      <Button role="tertiary">Tertiary</Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe("Accept Button", () => {
  test("secondary success Button", () => {
    const component = renderer.create(
      <Button appearance="success">Secondary</Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("primary success Button", () => {
    const component = renderer.create(
      <Button appearance="success" role="primary">
        Primary
      </Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("tertiary success Button", () => {
    const component = renderer.create(
      <Button appearance="success" role="tertiary">
        Primary
      </Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe("Disabled Button", () => {
  test("secondary disabled Button", () => {
    const component = renderer.create(<Button disabled>Secondary</Button>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("primary disabled Button", () => {
    const component = renderer.create(
      <Button disabled role="primary">
        Primary
      </Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("tertiary disabled Button", () => {
    const component = renderer.create(
      <Button disabled role="tertiary">
        Primary
      </Button>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
