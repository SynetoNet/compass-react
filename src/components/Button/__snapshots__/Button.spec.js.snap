// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accept Button primary success Button 1`] = `
<button
  className="btn-success width-auto btn btn-primary"
  color="success"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Primary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Accept Button secondary success Button 1`] = `
<button
  className="btn-success width-auto btn btn-primary"
  color="success"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Secondary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Accept Button tertiary success Button 1`] = `
<button
  className="btn-success width-auto btn btn-tertiary"
  color="success"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Primary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Default Button primary Button 1`] = `
<button
  className="btn-primary width-auto btn btn-primary"
  color="primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Primary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Default Button secondary Button 1`] = `
<button
  className="btn-primary width-auto btn btn-primary"
  color="primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Secondary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Default Button tertiary Button 1`] = `
<button
  className="btn-primary width-auto btn btn-tertiary"
  color="primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Tertiary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Disabled Button primary disabled Button 1`] = `
<button
  className="btn-primary width-auto disabled btn btn-primary"
  color="primary"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Primary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Disabled Button secondary disabled Button 1`] = `
<button
  className="btn-primary width-auto disabled btn btn-primary"
  color="primary"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Secondary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;

exports[`Disabled Button tertiary disabled Button 1`] = `
<button
  className="btn-primary width-auto disabled btn btn-tertiary"
  color="primary"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Primary
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</button>
`;
